### PR TITLE
Add Jacobi-fused SpGEMM

### PR DESCRIPTION
**Pull request by seheracer**
**Description:**

This PR adds support for Jacobi-fused SpGEMM in the form of C = (I - w D^-1 A)B.
_Inputs:_
_A:_ nxn sparse matrix with nonzero diagonal entries.
_B:_ nxm sparse matrix
_w:_ scalar
_D^-1:_ nx1 2D view.

For the people who are familiar with KK SpGEMM:
- Jacobi-SpGEMM uses the existing spgemm_symbolic as is, because the nonzero pattern of A * B and (I - w D^-1 A) * B are the same as long as all diagonal entries of A have nonzero values. 
-  Most of the newly-added implementation files, whose names contain spgemm_jacobi, are updated versions of their spgemm_numeric counterparts. These updates incorporate w and D^-1 to the computation properly. 

**Testing:**

```
[sacer@white28 kokkos-kernels]$ scripts/cm_test_all_sandia --spot-check --arch=Power8,Pascal60
Running on machine: white
WARNING!! THE FOLLOWING CHANGES ARE UNCOMMITTED!! :
M  src/CMakeLists.txt
A  src/impl/generated_specializations_cpp/spgemm_jacobi/KokkosSparse_spgemm_jacobi_eti_spec_inst.cpp.in
A  src/impl/generated_specializations_hpp/KokkosSparse_spgemm_jacobi_eti_spec_avail.hpp.in
A  src/impl/generated_specializations_hpp/KokkosSparse_spgemm_jacobi_eti_spec_decl.hpp.in
M  src/impl/tpls/KokkosSparse_spgemm_tpl_spec_avail.hpp
M  src/sparse/KokkosSparse_spgemm.hpp
A  src/sparse/KokkosSparse_spgemm_jacobi.hpp
M  src/sparse/impl/KokkosSparse_spgemm_impl.hpp
AM src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp
AM src/sparse/impl/KokkosSparse_spgemm_jacobi_seq_impl.hpp
AM src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp
A  src/sparse/impl/KokkosSparse_spgemm_jacobi_spec.hpp
A  unit_test/cuda/Test_Cuda_Sparse_spgemm_jacobi.cpp
A  unit_test/openmp/Test_OpenMP_Sparse_spgemm_jacobi.cpp
A  unit_test/serial/Test_Serial_Sparse_spgemm_jacobi.cpp
A  unit_test/threads/Test_Threads_Sparse_spgemm_jacobi.cpp
?? unit_test/sparse/Test_Sparse_spgemm_jacobi.hpp

KokkosKernels Repository Status:  27712a948276f8525fe4db0f7f0f8357272fa55a Merge pull request #681 from seheracer/diagDominantFix

Kokkos Repository Status:  b5b7660af56d883ffc8892fa8a253ee257f38e26 Merge pull request #2800 from crtrott/fix-update-copyright-master


Going to test compilers:  gcc/6.4.0 gcc/7.2.0 ibm/16.1.0 cuda/9.2.88 cuda/10.1.105
Testing compiler gcc/6.4.0
Testing compiler gcc/7.2.0
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
  Starting job gcc-6.4.0-OpenMP_Serial-release
kokkos options:
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
kokkos options:
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
kokkos options:
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-7.2.0-Serial-release
Testing compiler ibm/16.1.0
  Starting job gcc-7.2.0-OpenMP_Serial-release
kokkos options:
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-7.2.0-OpenMP_Serial-release
Testing compiler cuda/9.2.88
Unrecognized compiler ibm/16.1.0 when looking for Spack variants
Unrecognized compiler ibm/16.1.0 when looking for Spack variants
Unrecognized compiler ibm/16.1.0 when looking for Spack variants
  Starting job ibm-16.1.0-Serial-release
kokkos options:
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED ibm-16.1.0-Serial-release
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
  Starting job cuda-9.2.88-Cuda_OpenMP-release
kokkos options:
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler cuda/10.1.105
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
  Starting job cuda-9.2.88-Cuda_Serial-release
kokkos options:
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-9.2.88-Cuda_Serial-release
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
  Starting job cuda-10.1.105-Cuda_OpenMP-release
kokkos options:
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-10.1.105-Cuda_OpenMP-release
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
  Starting job cuda-10.1.105-Cuda_Serial-release
kokkos options:
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-10.1.105-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.1.105-Cuda_OpenMP-release build_time=372 run_time=108
cuda-10.1.105-Cuda_Serial-release build_time=319 run_time=148
cuda-9.2.88-Cuda_OpenMP-release build_time=311 run_time=118
cuda-9.2.88-Cuda_Serial-release build_time=325 run_time=159
gcc-6.4.0-OpenMP_Serial-release build_time=136 run_time=130
gcc-7.2.0-OpenMP-release build_time=104 run_time=44
gcc-7.2.0-OpenMP_Serial-release build_time=122 run_time=128
gcc-7.2.0-Serial-release build_time=91 run_time=80
ibm-16.1.0-Serial-release build_time=604 run_time=88
```

### CODE REVIEW

**Comment by srajama1** [comment]
Before starting the review, a simple question - how does the fused version compare to non-fused version in terms of performance ?

**Comment by seheracer** [comment]
> Before starting the review, a simple question - how does the fused version compare to non-fused version in terms of performance ?

There is no kokkos-kernels way of comparing these performances, because kk does not have a non-fused version of jacobi-spgemm. Do you like this comparison in the Tpetra level using one GPU?

**Comment by srajama1** [comment]
The work in Jacobi is so small that it would be basically launch cost, right ?

**Comment by seheracer** [comment]
> The work in Jacobi is so small that it would be basically launch cost, right ?

I guess so. Not sure though. 

I made a comparison in the Tpetra level against the existing Tpetra MSAK algorithm  (https://github.com/trilinos/Trilinos/blob/67064dc0bd94754b56ded6ee2a0a09a2dcc45433/packages/tpetra/core/ext/TpetraExt_MatrixMatrix_ExtraKernels_def.hpp#L965)

Timings on 1  GPU (MSAK time -> Jacobi-fused SpGEMM time):
```
mac_econ_fwd500: 0.0232098 -> 0.0196707 
scircuit: 0.026749 -> 0.0251468 
hood: 0.0220142 -> 0.0201918 
pwtk: 0.0220315 -> 0.0203125 
coPaperCiteseer: 0.0976187 -> 0.0921416
```

Timings on 6 GPUs:
```
mac_econ_fwd500: 0.0811009 -> 0.060508
scircuit: 0.115992 -> 0.103049
hood: 0.151943 -> 0.0782865 
pwtk: 0.0777757 -> 0.0622603 
coPaperCiteseer: 0.228263 -> 0.203596
```

**Comment by srajama1** [comment]
Ok, these times look good. I will review this tonight.

**Comment by seheracer** [comment]
> Ok, these times look good. I will review this tonight.

Thanks. I updated the results of the multiple-GPU runs (from 6 GPUs to 4 GPUs), because I was on a single node on white so I should have used 4 GPUs instead of 6.

File: src/impl/generated_specializations_cpp/spgemm_jacobi/KokkosSparse_spgemm_jacobi_eti_spec_inst.cpp.in [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
```
Comment (srajama1) [inline src/impl/generated_specializations_cpp/spgemm_jacobi/KokkosSparse_spgemm_jacobi_eti_spec_inst.cpp.in:38]: I don't have to be contact for everything. Use your name.

File: src/sparse/KokkosSparse_spgemm_jacobi.hpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+		     typename bscalar_nnz_view_t_::const_value_type>::value,
+		     "KokkosSparse::spgemm_jacobi: scalar type of right handside matrix should be same as kernelHandle scalar.");
+
+      static_assert (std::is_same<typename KernelHandle::const_nnz_scalar_t,
+		     typename cscalar_nnz_view_t_::const_value_type>::value,
+		     "KokkosSparse::spgemm_jacobi: scalar type of output matrix should be same as kernelHandle scalar.");
+
+
+      if (transposeA || transposeB){
+	throw std::runtime_error ("SpGEMM is not implemented for Transposes yet. "
```
Comment (srajama1) [inline src/sparse/KokkosSparse_spgemm_jacobi.hpp:141]: You mean spgemm-jacobi doesn't support transpose ..

File: src/sparse/KokkosSparse_spgemm_jacobi.hpp [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+				  "If you need this case please let kokkos-kernels developers know.\n");
+      }
+
+
+      typedef typename KernelHandle::const_size_type c_size_t;
+      typedef typename KernelHandle::const_nnz_lno_t c_lno_t;
+      typedef typename KernelHandle::const_nnz_scalar_t c_scalar_t;
+
+      typedef typename KernelHandle::HandleExecSpace c_exec_t;
+      typedef typename KernelHandle::HandleTempMemorySpace c_temp_t;
```
Comment (srajama1) [inline src/sparse/KokkosSparse_spgemm_jacobi.hpp:151]: Do we actually use TempMemorySpace and PersistentMemorySpace ?

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+	switch (my_exec_space){
+	default:
+	  return row_index;
+#if defined( KOKKOS_ENABLE_SERIAL )
+	case KokkosKernels::Impl::Exec_SERIAL:
+	  return 0;
+#endif
+#if defined( KOKKOS_ENABLE_OPENMP )
+	case KokkosKernels::Impl::Exec_OMP:
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE
```
Comment (srajama1) [inline src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp:120]: It is better if our new codes shouldn't. depend on deprecated code, can we remove that ?

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+	  return Kokkos::OpenMP::hardware_thread_id();
+#else
+	  return Kokkos::OpenMP::impl_hardware_thread_id();
+#endif
+#endif
+#if defined( KOKKOS_ENABLE_THREADS )
+	case KokkosKernels::Impl::Exec_PTHREADS:
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE
```
Comment (srajama1) [inline src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp:128]: Remove deprecated code ?

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_seq_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  typename alno_nnz_view_t_,
+  typename ascalar_nnz_view_t_,
+  typename blno_row_view_t_,
+  typename blno_nnz_view_t_,
+  typename bscalar_nnz_view_t_,
+  typename clno_row_view_t_,
+  typename clno_nnz_view_t_,
+  typename cscalar_nnz_view_t_,
+  typename dinv_scalar_view_t>
+void spgemm_debug_jacobi(
```
Comment (srajama1) [inline src/sparse/impl/KokkosSparse_spgemm_jacobi_seq_impl.hpp:62]: Debug in the name .. can we remove it ?

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,1542 @@
...
+	switch (my_exec_space){
+	default:
+	  return row_index;
+#if defined( KOKKOS_ENABLE_SERIAL )
+	case KokkosKernels::Impl::Exec_SERIAL:
+	  return 0;
+#endif
+#if defined( KOKKOS_ENABLE_OPENMP )
+	case KokkosKernels::Impl::Exec_OMP:
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE
```
Comment (srajama1) [inline src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp:209]: deprecated code

File: unit_test/threads/Test_Threads_Sparse_spgemm_jacobi.cpp [hunk 8]
```diff
@@ -0,0 +1,2 @@
+#include<Test_Threads.hpp>
```
Comment (srajama1) [inline unit_test/threads/Test_Threads_Sparse_spgemm_jacobi.cpp:1]: Can we also add a perf-test or add an option to spgemm perf-test ?

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,1542 @@
+/*
+//@HEADER
```
Comment (srajama1) [inline src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp:2]: I was looking for a case where d_inv will be read once for the entire team when all threads in the team are working on a row, couldn't see it. May be you haven't considered that level of optimization.

File: src/sparse/KokkosSparse_spgemm_jacobi.hpp (hunk 3 above)
Comment (seheracer) [reply src/sparse/KokkosSparse_spgemm_jacobi.hpp:151]: I am not able to tell if this is absolutely required or not. TempMemorySpace and PersistentMemorySpace go into the const_handle_type, which is used in the KokkosSparse::Impl::SPGEMM_JACOBI call at the end of this file.

Comment (seheracer) [reply src/sparse/KokkosSparse_spgemm_jacobi.hpp:151]: I mean, this is a copy of the file KokkosSparse_spgemm_numeric.hpp which is updated for jacobi accordingly.

Comment (srajama1) [reply src/sparse/KokkosSparse_spgemm_jacobi.hpp:151]: Understood, this requires a cleanup in spgemm as well. I think the handle might have reasonable defaults. If you think it is hard to remove we can come to it w/ clean up pass for both codes later.

Comment (seheracer) [reply src/sparse/KokkosSparse_spgemm_jacobi.hpp:151]: Yes, it is hard to remove at the moment. Let's do it later.

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp (hunk 9 above)
Comment (seheracer) [reply src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp:2]: I think this line is already doing what you ask for: https://github.com/seheracer/kokkos-kernels/blob/247a195060fc1e167e38a8a62ce3a7f34c2515c9/src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp#L462

For all threads who work on the row with id row_index, dinv(row_index,0) is read once. No?

File: src/impl/generated_specializations_cpp/spgemm_jacobi/KokkosSparse_spgemm_jacobi_eti_spec_inst.cpp.in (hunk 1 above)
Comment (seheracer) [reply src/impl/generated_specializations_cpp/spgemm_jacobi/KokkosSparse_spgemm_jacobi_eti_spec_inst.cpp.in]: Done.

File: src/sparse/KokkosSparse_spgemm_jacobi.hpp (hunk 2 above)
Comment (seheracer) [reply src/sparse/KokkosSparse_spgemm_jacobi.hpp]: Done.

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp (hunk 4 above)
Comment (seheracer) [reply src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp]: Done.

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp (hunk 5 above)
Comment (seheracer) [reply src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp]: Done.

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_seq_impl.hpp (hunk 6 above)
Comment (seheracer) [reply src/sparse/impl/KokkosSparse_spgemm_jacobi_seq_impl.hpp]: Changed:
spgemm_debug_jacobi -> spgemm_jacobi_seq
SPGEMM_DEBUG -> SPGEMM_SERIAL

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp (hunk 7 above)
Comment (seheracer) [reply src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp]: Done.

File: unit_test/threads/Test_Threads_Sparse_spgemm_jacobi.cpp (hunk 8 above)
Comment (seheracer) [reply unit_test/threads/Test_Threads_Sparse_spgemm_jacobi.cpp:1]: I added the perf test separately.

**Comment by seheracer** [comment]
@srajama1 I did the changes you asked for. The new spot-check output:


```
[sacer@white28 kokkos-kernels]$ scripts/cm_test_all_sandia --spot-check --arch=Power8,Pascal60
Running on machine: white

KokkosKernels Repository Status:  570be02d8cad38be706802f518b9f619f9455da5 Change contact

Kokkos Repository Status:  b5b7660af56d883ffc8892fa8a253ee257f38e26 Merge pull request #2800 from crtrott/fix-update-copyright-master


Going to test compilers:  gcc/6.4.0 gcc/7.2.0 ibm/16.1.0 cuda/9.2.88 cuda/10.1.105
Testing compiler gcc/6.4.0
Testing compiler gcc/7.2.0
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
Unrecognized compiler gcc/6.4.0 when looking for Spack variants
  Starting job gcc-6.4.0-OpenMP_Serial-release
kokkos options:
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
kokkos options:
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
kokkos options:
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-7.2.0-Serial-release
Testing compiler ibm/16.1.0
  Starting job gcc-7.2.0-OpenMP_Serial-release
kokkos options:
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
  PASSED gcc-7.2.0-OpenMP_Serial-release
Testing compiler cuda/9.2.88
Unrecognized compiler ibm/16.1.0 when looking for Spack variants
Unrecognized compiler ibm/16.1.0 when looking for Spack variants
Unrecognized compiler ibm/16.1.0 when looking for Spack variants
  Starting job ibm-16.1.0-Serial-release
kokkos options:
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED ibm-16.1.0-Serial-release
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
  Starting job cuda-9.2.88-Cuda_OpenMP-release
kokkos options:
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler cuda/10.1.105
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
Unrecognized compiler cuda/9.2.88 when looking for Spack variants
  Starting job cuda-9.2.88-Cuda_Serial-release
kokkos options:
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-9.2.88-Cuda_Serial-release
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
  Starting job cuda-10.1.105-Cuda_OpenMP-release
kokkos options:
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-10.1.105-Cuda_OpenMP-release
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
Unrecognized compiler cuda/10.1.105 when looking for Spack variants
  Starting job cuda-10.1.105-Cuda_Serial-release
kokkos options:
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
  PASSED cuda-10.1.105-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.1.105-Cuda_OpenMP-release build_time=349 run_time=107
cuda-10.1.105-Cuda_Serial-release build_time=315 run_time=148
cuda-9.2.88-Cuda_OpenMP-release build_time=361 run_time=119
cuda-9.2.88-Cuda_Serial-release build_time=370 run_time=159
gcc-6.4.0-OpenMP_Serial-release build_time=144 run_time=129
gcc-7.2.0-OpenMP-release build_time=96 run_time=44
gcc-7.2.0-OpenMP_Serial-release build_time=138 run_time=127
gcc-7.2.0-Serial-release build_time=92 run_time=79
ibm-16.1.0-Serial-release build_time=622 run_time=88 
```

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,1534 @@
...
+	KokkosKernels::Experimental::HashmapAccumulator<nnz_lno_t,nnz_lno_t,scalar_t>
+	  hm(thread_shmem_hash_size, thread_shmem_key_size, begins, nexts, keys, vals);
+
+	KokkosKernels::Experimental::HashmapAccumulator<nnz_lno_t,nnz_lno_t,scalar_t>
+	  hm2(pow2_hash_size, pow2_hash_size,
+	      NULL, NULL, NULL, NULL);
+
+	Kokkos::parallel_for(Kokkos::TeamThreadRange(teamMember, team_row_begin, team_row_end), [&] (const nnz_lno_t& row_index) {
+
+	    scalar_t mult = -omega*dinv(row_index,0);
```
Comment (brian-kelley) [inline src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp:454]: @srajama1 I agree with @seheracer, each row_index is handled exactly once and this is a warp-broadcast access of dinv, so it's already optimal (copying into shared first also accesses the same global addresses once).

File: src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,1534 @@
...
+	  });
+
+	int bs = vector_size * suggested_team_size;
+	int vector_shift = thread_rank * vector_size + vector_rank;
+
+	for (nnz_lno_t row_index = team_row_begin; row_index < team_row_end; ++row_index){
+
+	  teamMember.team_barrier();
+
+	  scalar_t mult = -omega*dinv(row_index,0);
```
Comment (brian-kelley) [inline src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp:810]: In this case, it's a whole team accessing the same address for dinv so it's not quite optimal in bandwidth terms, but I doubt there's a win from copying to shared first.

**Comment by srajama1** [comment]
Thanks, @brian-kelley ! I will merge it. 

@seheracer : We need to add some documentation for this in the private github.

**Pull request merged**